beacon-chain/blockchain/core.go

```diff
@@ -319,59 +322,98 @@ func (b *BeaconChain) saveBlock(block *types.Block) error {
 	return b.db.Put(hash[:], encodedState)
 }
 
-// processAttestations processes the attestations of an incoming block.
-func (b *BeaconChain) processAttestations(block *types.Block) error {
+// processAttestation processes the attestation of an incoming block.
+func (b *BeaconChain) processAttestation(attestation *pb.AttestationRecord, block *types.Block) error {
 	// Validate attestation's slot number has is within range of incoming block number.
 	slotNumber := int(block.SlotNumber())
-	for _, attestation := range block.Attestations() {
-		if int(attestation.Slot) > slotNumber {
-			return fmt.Errorf("attestation slot number can't be higher than block slot number. Found: %v, Needed lower than: %v",
-				attestation.Slot,
-				slotNumber)
-		}
-		if int(attestation.Slot) < slotNumber-params.CycleLength {
-			return fmt.Errorf("attestation slot number can't be lower than block slot number by one CycleLength. Found: %v, Needed greater than: %v",
-				attestation.Slot,
-				slotNumber-params.CycleLength)
-		}
+	if int(attestation.Slot) > slotNumber {
+		return fmt.Errorf("attestation slot number can't be higher than block slot number. Found: %v, Needed lower than: %v",
+			attestation.Slot,
+			slotNumber)
+	}
+	if int(attestation.Slot) < slotNumber-params.CycleLength {
+		return fmt.Errorf("attestation slot number can't be lower than block slot number by one CycleLength. Found: %v, Needed greater than: %v",
+			attestation.Slot,
+			slotNumber-params.CycleLength)
+	}
 
-		// Get all the block hashes up to cycle length.
-		parentHashes := b.getSignedParentHashes(block, attestation)
-		attesterIndices, err := b.getAttesterIndices(attestation)
-		if err != nil {
-			return err
-		}
+	// Get all the block hashes up to cycle length.
+	parentHashes := b.getSignedParentHashes(block, attestation)
+	attesterIndices, err := b.getAttesterIndices(attestation)
+	if err != nil {
+		return err
```

Comment:
More verbose error messages here with fmt.Errorf. I can see this method failing in a variety of situations and will make life easier down the line.